#  Part 1.  Computing with Python 3

Welcome to programming!

What is the difference between *Python* and a calculator?  We begin this first lesson by showing how Python can be used **as** a calculator, and we move into one of the most important programming structures -- the **loop**.  Loops allow computers to carry out repetetive computations, with just a few commands.

## Table of Contents

- [Python as a calculator](#calculator)
- [Calculating with booleans](#booleans)
- [Declaring variables](#variables)
- [Ranges](#ranges)
- [Iterating over a range](#iterating)
- [Explorations](#explore)

<a id='calculator'></a>

## Python as a calculator

Different kinds of data are stored as different *types* in Python.  For example, if you wish to work with integers, your data is typically stored as an *int*.  A real number might be stored as a *float*.  There are types for booleans (True/False data), strings (like "Hello World!"), and many more we will see.  

A more complete reference for Python's numerical types and arithmetic operations can be found in the [official Python documentation](https://docs.python.org/3/library/stdtypes.html).  The [official Python tutorial](https://docs.python.org/3/tutorial/introduction.html) is also a great place to start.

Python allows you to perform arithmetic operations:  addition, subtraction, multiplication, and division, on numerical types.  The operation symbols are `+`, `-`, `*`, and `/`.   Evaluate each of the following cells to see how Python performs operations on *integers*.  To evaluate the cell, click anywhere within the cell to select it (a selected cell will probably have a thick <span style="color:green">green</span> line on its left side) and use the keyboard shortcut *Shift-Enter* to evaluate.  As you go through this and later lessons, try to *predict* what will happen when you evaluate the cell before you hit Shift-Enter.  

In [1]:
2 + 3

5

In [2]:
2 * 3

6

In [3]:
5 - 11

-6

In [4]:
5.0 - 11

-6.0

In [5]:
5 / 11

0.45454545454545453

In [6]:
6 / 3

2.0

In [7]:
5 // 11

0

In [8]:
6 // 3

2

The results are probably not too surprising, though the last two require a bit of explanation.  Python *interprets* the input number 5 as an *int* (integer) and 5.0 as a *float*.  "Float" stands for "floating point number," which are decimal approximations to real numbers.  The word "float" refers to the fact that the decimal (or binary, for computers) point can float around (as in 1.2345 or 12.345 or 123.45 or 1234.5 or 0.00012345).  There are deep computational issues related to how computers handle decimal approximations, and you can [read about the IEEE standards](https://en.wikipedia.org/wiki/IEEE_754) if you're interested.

Python enables different kinds of division.  The single-slash division in Python 3.x gives a floating point approximation of the quotient.  That's why `5 / 11` and `6 / 3` both output floats.  On the other hand, `5 // 11` and `6 // 3` yield integer outputs (rounding down) -- this is useful, but one has to be careful!

In fact the designers of Python changed their mind.  **This tutorial assumes that you are using Python 3.x.**  If you are using Python 2.x, the command `5 / 11` would output zero.

In [9]:
-12 // 5  # What will this output?  Guess before evaluating!

-3

**Note**: -12/5 = -2.4, so rounding *down* gives -3

Why use integer division `//` and why use floating point division?  In practice, integer division is typically a faster operation.  So if you only need the rounded result (and that will often be the case), use integer division.  It will run much faster than carrying out floating point division then manually rounding down.

Observe that floating point operations involve approximation.  The result of `5.0/11.0` might not be what you expect in the last digit.  Over time, especially with repeated operations, *floating point approximation* errors can add up!

You might be wondering about the little In[XX] and Out[XX] prompts.  What is their purpose?  Guess what the following line will do.

In [11]:
Out[1] * Out[2]

30

Cool, huh?  It's nice to have a record of previous computations, especially if you don't want to type something again.

Python allows you to group expressions with parentheses, and follows the order of operations that you learn in school.

In [12]:
(3 + 4) * 5

35

In [13]:
3 + (4 * 5)

23

In [14]:
3 + 4 * 5   #  What do you think will be the result?  Remember PEMDAS?

23

Now is a good time to try a few computations of your own, in the empty cell below.  You can type any Python commands you want in the empty cell.  If you want to insert a new cell into this notebook, it takes two steps:
1.  Click **to the left** of any existing cell.  This should make a <span style="color:blue">blue</span> bar appear to the left of the cell.
2.  Use the keyboard shortcut **a** to insert a new cell **above** the blue-selected cell or **b** to insert a new cell **below** the blue-selected cell.
You can also use the keyboard shortcut **x** do delete a blue-selected cell... be careful!

In [15]:
#  An empty cell.  Have fun!
(2.87351691727)*(4.1128569993238472118935672)

11.818364165869404

In [16]:
2.8*4.1

11.479999999999999

In [17]:
Out[15] - Out[16]

0.33836416586940565

In [18]:
(0.07351691727)*(4.1128569993238472118935672) + (0.0128569993238472118935672)*(2.87351691727) + (0.07351691727)*(0.0128569993238472118935672)

0.3402545797806673

Oh no where'd the rest go???

For number theory, *division with remainder* is an operation of central importance.  Integer division provides the quotient, and the operation `%` provides the remainder.  It's a bit strange that the percent symbol is used for the remainder, but this [dates at least to the early 1970s](https://softwareengineering.stackexchange.com/questions/294297/in-what-programming-language-did-the-use-of-the-percent-sign-to-mean-modulo) and has become standard across computer languages.

In [19]:
23 // 5  # Integer division

4

In [20]:
23 % 5  # The remainder after division

3

Note in the code above, there are little "comments".  To place a short comment on a line of code, just put a hashtag `#` at the end of the line of code, followed by your comment.

Python gives a single command for division with remainder.  Its output is a *tuple*.

In [21]:
divmod(23,5)

(4, 3)

In [22]:
type(divmod(23,5))

tuple

All data in Python has a type, but a common complaint about Python is that types are a bit concealed "under the hood".  But they are not far under the hood!  Anyone can find out the type of some data with a single command.

In [23]:
type(3)

int

In [24]:
type(3.0)

float

In [25]:
type('Hello')

str

In [26]:
type([1,2,3])

list

The key to careful computation in Python is always being *aware of the type* of your data, and *knowing* how Python operates differently on data of different types.

In [27]:
3 + 3

6

In [28]:
3.0 + 3.0

6.0

In [29]:
'Hello' + 'World!'

'HelloWorld!'

In [30]:
[1,2,3] + [4,5,6]

[1, 2, 3, 4, 5, 6]

In [31]:
3 + 3.0

6.0

In [32]:
3 + 'Hello!'  # Uh oh!

TypeError: unsupported operand type(s) for +: 'int' and 'str'

In [33]:
#  An empty cell.  Have fun!
#  Try operating on ints, floats, and strings, with different operations.  Which ones work?  How?

str(3) + ' Hello!'

'3 Hello!'

In [35]:
type(str(3))

str

In [36]:
type(3/4)

float

In [37]:
type(3//4)

int

In [40]:
int(5.55555) # Floors apparently

5

As you can see, addition (the `+` operator) is interpreted differently in the contexts of numbers, strings, and lists.  The designers of Python allowed us to add *numbers* of different types:  if you try to operate on an *int* and a *float*, the *int* will typically be *coerced* into a float in order to perform the operation.  But the designers of Python did not give meaning to the addition of a number with a string, for example.  That's why you probably received a *TypeError* after trying to add a number to a string. 

On the other hand, Python does interpret *multiplication* of a natural number with a string or a list.

In [41]:
3 * 'Hello!'

'Hello!Hello!Hello!'

In [42]:
0 * 'Hello!'

''

In [43]:
2 * [1,2,3]

[1, 2, 3, 1, 2, 3]

Can you create a string with 100 A's (like `AAA...`)?  Use an appropriate operation in the cell below.

In [44]:
#  Practice cell
100*"A"

'AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA'

Exponents in Python are given by the `**` operator.  The following lines compute 2 to the 1000th power, in two different ways.

In [45]:
2**1000

10715086071862673209484250490600018105614048117055336074437503883703510511249361224931983788156958581275946729175531468251871452856923140435984577574698574803934567774824230985421074605062371141877954182153046474983581941267398767559165543946077062914571196477686542167660429831652624386837205668069376

In [46]:
2.0**1000

1.0715086071862673e+301

As before, Python interprets an operation (`**`) differently in different contexts.  When given integer input, Python evaluates `2**1000` **exactly**.  The result is a large integer.  A nice fact about Python, for mathematicians, is that it handles exact integers of arbitrary length!  Many other programming languages (like C++) will give an error message if integers get too large in the midst of a computation.  

New in version 3.x, Python implements long integers without giving signals to the programmer or changing types.  In Python 2.x, there were two types: *int* for somewhat small integers (e.g., up to $2^{31}$) and *long* type for all larger integers.  Python 2.x would signal which type of integer was being used, by placing the letter "L" at the end of a long integer.  Now, in Python 3.x, the programmer doesn't really see the difference.  There is only the *int* type.  But Python still optimizes computations, using hardware functionality for arithmetic of small integers and custom routines for large integers.  The programmer doesn't have to worry about it most of the time.

For scientific applications, one often wants to keep track of only a certain number of significant digits (sig figs).  If one computes the floating point exponent `2.0**1000`, the result is a decimal approximation.  It is still a float.  The expression "e+301" stands for "multiplied by 10 to the 301st power", i.e., Python uses *scientific notation* for large floats.

In [47]:
type(2**1000)

int

In [48]:
type(2.0**1000)

float

In [49]:
#  An empty cell.  Have fun!
"A"**7

TypeError: unsupported operand type(s) for ** or pow(): 'str' and 'int'

In [50]:
int(5.888)

5

In [51]:
int("A")

ValueError: invalid literal for int() with base 10: 'A'

Now is a good time for reflection.  Double-click in the cell below to answer the given questions.  Cells like this one are used for text rather than Python code.  Text is entered using *markdown*, but you can typically just enter text as you would in any text editor without problems.  Press *shift-Enter* after editing a markdown cell to complete the editing process.  

Note that a dropdown menu in the toolbar above the notebook allows you to choose whether a cell is Markdown or Code (or a few other things), if you want to add or remove markdown/code cells.

### Exercises

1.  What data types have you seen, and what kinds of data are they used for?  Can you remember them without looking back?

2.  How is division `/` interpreted differently for different types of data?

3.  How is multiplication `*` interpreted differently for different types of data?

4.  What is the difference between 100 and 100.0, for Python?

1. ints, floats, strings, tuples, lists


2. It looks like / takes combinations of integers and floats and always give a float, even if the result is, mathematically, an integer. The // division performs division and rounds *down.* (Question: is rounding down the same as 'flooring?' I.e. is 'floor'(-2.6) = -2 or -3?) Unlike /, we also have that // will result in an integer if the two arguments are integers. If one argument is a float, the result will also be a float, even if the decimal is x.0.    


3. Two ints multiplied yeilds an int. If there is a float present, the result will be a float, even if the result is in $\mathbb{Z}$. Strings multiplied by an integer $n$ concatenate the string to itself $n$ times. A list times $n$ appends the list to itself $n$ times.


4. 100 is an int, 100.0 is a float.

#### Tests for the above questions:

In [57]:
6/2

3.0

In [52]:
2.345//2

1.0

In [56]:
2//2

1

In [53]:
"thirteen"/"2"

TypeError: unsupported operand type(s) for /: 'str' and 'str'

In [54]:
[1,2,3]/[4,5,6]

TypeError: unsupported operand type(s) for /: 'list' and 'list'

In [55]:
(8,4)/2

TypeError: unsupported operand type(s) for /: 'tuple' and 'int'

In [59]:
2*3

6

<a id='booleans'></a>

In [60]:
2.0*3

6.0

In [61]:
[1,2,3]*6

[1, 2, 3, 1, 2, 3, 1, 2, 3, 1, 2, 3, 1, 2, 3, 1, 2, 3]

## Calculating with booleans

A *boolean* (type *bool*) is the smallest possible piece of data.  While an *int* can be any integer, positive or negative, a *boolean* can only be one of two things:  *True* or *False*.  In this way, booleans are useful for storing the answers to yes/no questions.  

Questions about (in)equality of numbers are answered in Python by *operations* with numerical input and boolean output.  Here are some examples.  A more complete reference is [in the official Python documentation](https://docs.python.org/3/library/stdtypes.html#boolean-operations-and-or-not).

In [62]:
3 > 2

True

In [63]:
type(3 > 2)

bool

In [64]:
10 < 3

False

In [65]:
2.4 < 2.4000001

True

In [66]:
32 >= 32

True

In [67]:
32 >= 31

True

In [68]:
2 + 2 == 4

True

Which number is bigger:  $23^{32}$ or $32^{23}$?  Use the cell below to answer the question!

In [75]:
#  23**32 is larger, and not equal to either
23**32 > 32*23

True

The expressions `<`, `>`, `<=`, `>=` are interpreted here as **operations** with numerical input and boolean output.  The symbol `==` (two equal symbols!) gives a True result if the numbers are equal, and False if the numbers are not equal.  An extremely common typo is to confuse `=` with `==`.  But the single equality symbol `=` has an entirely different meaning, as we shall see.

Using the remainder operator `%` and equality, we obtain a divisibility test.

In [76]:
63 % 7 == 0  # Is 63 divisible by 7?

True

In [77]:
101 % 2 == 0  # Is 101 even?

False

Use the cell below to determine whether 1234567890 is divisible by 3.

In [78]:
# Your code goes here.
1234567890 % 3 == 0

True

Booleans can be operated on by the standard logical operations: and, or, not.  In ordinary English usage, "and" and "or" are conjunctions, while here in *Boolean algebra*, "and" and "or" are operations with Boolean inputs and Boolean output.  The precise meanings of "and" and "or" are given by the following **truth tables**.

    
| and | True | False |
|-----|------|-------|
| **True** | True | False |
| **False** | False | False|
  
| or | True | False |
|-----|------|-------|
| **True** | True | True |
| **False** | True | False|

In [79]:
True and False

False

In [80]:
True or False

True

In [81]:
True or True

True

In [82]:
not True

False

Use the truth tables to predict the result (True or False) of each of the following, before evaluating the code.

In [83]:
(2 > 3) and (3 > 2)

False

In [84]:
(1 + 1 == 2) or (1 + 1 == 3)

True

In [85]:
not (-1 + 1 >= 0)

False

In [86]:
2 + 2 == 4

True

In [87]:
2 + 2 != 4  # For "not equal", Python uses the operation `!=`.

False

In [88]:
2 + 2 != 5  # Is 2+2 *not* equal to 5?

True

In [89]:
not (2 + 2 == 5)  # The same as above, but a bit longer to write.

True

Experiment below to see how Python handles a double or triple negative, i.e., something with a `not` `not`.

In [90]:
# Experiment here.
not not True

True

In [91]:
not not not not not not not not (1 != 2)

True

Python does give an interpretation to arithmetic operations with booleans and numbers.  Try to guess this interpretation with the following examples.  Change the examples to experiment!

In [92]:
False * 100

0

In [93]:
True + 13

14

In [95]:
type(True + True)

int

#### I'm proud of this one:

In [96]:
(2+2 != 5) + (2 + 2) == 5

True

This ability of Python to interpret operations based on context is a mixed blessing.  On one hand, it leads to handy shortcuts -- quick ways of writing complicated programs.  On the other hand, it can lead to code that is harder to read, especially for a Python novice.  Good programmers aim for code that is easy to read, not just short!

The [Zen of Python](https://www.python.org/dev/peps/pep-0020/) is a series of 20 aphorisms for Python programmers.  The first seven are below.

> Beautiful is better than ugly.

> Explicit is better than implicit.

> Simple is better than complex.

> Complex is better than complicated.

> Flat is better than nested.

> Sparse is better than dense.

> Readability counts.

### Exercises

1.  Did you look at the truth tables closely?  Can you remember, from memory, what `True or False` equals, or what `True and False` equals?  

2.  How might you easily remember the truth tables?  How do they resemble the standard English usage of the words "and" and "or"?

3.  If you wanted to know whether a number, like 2349872348723, is a multiple of 7 but **not** a multiple of 11, how might you write this in one line of Python code?

4.  You can chain together `and` commands, e.g., with an expression like `True and True and True` (which would evaluate to `True`).  You can also group booleans, e.g., with `True and (True or False)`.  Experiment to figure out the order of operations (`and`, `or`, `not`) for booleans.

6.  The operation `xor` means "exclusive or".  Its truth table is: `True xor True = False` and `False xor False = False` and `True xor False = True` and `False xor True = True`.  How might you implement `xor` in terms of the usual `and`, `or`, and `not`?



###  Solutions

1. I checked them out, yes. T or F = T; T and F = F.


2. `And` requires *both* argurments to be true. This is just like english. The `or` is inclusive, unlike english where it's usually interpreted as exlusive. 


3. `(2349872348723 % 7 == 0) and not (2349872348723 % 11 == 0)` will return true if the number is divisible by 7 but not 11.


4. The order seems to be `not, and, or`. The computation `False and True or True = True` shows that the grouping is `(False and True) or True`. The other ordering would yeild False. Further, `not` is evaluated before `and`, because `not False and False` evaluates to `False`. This occures only in the groupping `(not False) and False`.


5. For truth values $X$ and $Y$, $X$ `xor` $Y$ = ($X$ `or` $Y$) `and not` ($X$ `and` $Y$)

### Scratch

In [13]:
False and True or True

True

In [15]:
not False and False

False

<a id='variables'></a>

## Declaring variables

A central feature of programming is the declaration of variables.  When you declare a variable, you are *storing* data in the computer's *memory* and you are assigning a *name* to that data.  Both storage and name-assignment are carried out with the *single* equality symbol =.

In [16]:
e = 2.71828

With this command, the float 2.71828 is stored somewhere inside your computer, and Python can access this stored number by the name "e" thereafter.  So if you want to compute "e squared", a single command will do.

In [17]:
e * e

7.3890461584

In [18]:
type(e)

float

You can use just about any name you want for a variable, but your name *must* start with a letter, *must* not contain spaces, and your name *must* not be an existing Python word.  Characters in a variable name can include letters (uppercase and lowercase) and numbers and underscores `_`.  

So `e` is a valid name for a variable, but `type` is a bad name.  It is very tempting for beginners to use very short abbreviation-style names for variables (like `dx` or `vbn`).  But resist that temptation and use more descriptive names for variables, like `difference_x` or `very_big_number`.  This will make your code readable by you and others!

There are different style conventions for variable names.  We use lowercase names, with underscores separating words,  roughly following [Google's style conventions](https://google.github.io/styleguide/pyguide.html#Python_Style_Rules) for Python code.

In [23]:
my_number = 17

In [24]:
my_number < 23

True

After you declare a variable, its value remains the same until it is changed.  You can change the value of a variable with a simple assignment.  After the above lines, the value of my_number is 17.

In [25]:
my_number = 3.14

This command reassigns the value of my_number to 3.14.  Note that it changes the type too!  It effectively overrides the previous value and replaces it with the new value.

Often it is useful to change the value of a variable *incrementally* or *recursively*.  Python, like many programming languages, allows one to assign variables in a self-referential way.  What do you think the value of S will be after the following four lines?

In [26]:
S = 0
S = S + 1
S = S + 2
S = S + 3
print(S)

6


The first line `S = 0` is the initial declaration:  the value 0 is stored in memory, and the name S is assigned to this value.

The next line `S = S + 1` looks like nonsense, as an algebraic sentence.  But reading = as **assignment** rather than **equality**, you should read the line `S = S + 1` as assigning the *value* `S + 1` to the *name* `S`.  When Python interprets `S = S + 1`, it carries out the following steps.

1.  Compute the value of the right side, `S+1`.  (The value is 1, since `S` was assigned the value 0 in the previous line.)
2.  Assign this value to the left side, `S`.  (Now `S` has the value 1.)

Well, this is a slight lie.  Python probably does something more efficient, when given the command `S = S + 1`, since such operations are hard-wired in the computer and the Python interpreter is smart enough to take the most efficient route.  But at this level, it is most useful to think of a self-referential assignment of the form `X = expression(X)` as a two step process as above.

1.  Compute the value of `expression(X)`.
2.  Assign this value to `X`.

Now consider the following three commands.

In [27]:
my_number = 17
new_number = my_number + 1
my_number = 3.14

What are the values of the variables my_number and new_number, after the execution of these three lines?

To access these values, you can use the *print* function.

In [28]:
print(my_number)
print(new_number)

3.14
18


Python is an *interpreted* language, which carries out commands line-by-line from top to bottom.  So consider the three lines

``` python
my_number = 17
new_number = my_number + 1
my_number = 3.14
```

Line 1 sets the value of my_number to 17.  Line 2 sets the value of new_number to 18.  Line 3 sets the value of my_number to 3.14.  But Line 3 does *not* change the value of new_number at all.

(This will become confusing and complicated later, as we study mutable and immutable types.)

### Exercises

1.  What is the difference between `=` and `==` in the Python language?

2.  If the variable `x` has value `3`, and you then evaluate the Python command `x = x * x`, what will be the value of `x` after evaluation?

3.  Imagine you have two variables `a` and `b`, and you want to switch their values.  How could you do this in Python?

### Solutions

1. `=` *asigns* equality, while `==` compares it. `x = 2` *sets* x to 2, and then `x == 2` will return `true` because x *is* 2. 


2. 9


3. You'd have to set up a temporary variable. The code `a = b; b = a` would leave both a *and* b having the initial value of b. The better aproach would be: 
``` Python
old_a = a
a = b
b = old_a
```

<a id='ranges'></a>

## Lists and ranges

Python stands out for the central role played by *lists*.  A *list* is what it sounds like -- a list of data.  Data within a list can be of any type.  Multiple types are possible within the same list!  The basic syntax for a list is to use brackets to enclose the list items and commas to separate the list items.

In [29]:
type([1,2,3])

list

In [30]:
type(['Hello',17])

list

There is another type called a *tuple* that we will use less often.  Tuples use parentheses for enclosure instead of brackets.

In [31]:
type((1,2,3))

tuple

There's another list-like type in Python 3, called the `range` type.  Ranges are kind of like lists, but instead of plunking every item into a slot of memory, ranges just have to remember three integers:  their *start*, their *stop*, and their *step*.    

The `range` command creates a range with a given start, stop, and step.  If you only input one number, the range will ***start at zero*** and use ***steps of one*** and will stop ***just before*** the given stop-number.

One can create a list from a range (plunking every term in the range into a slot of memory), by using the `list` command.  Here are a few examples.

In [32]:
type(range(10)) # Ranges are their own type, in Python 3.x.  Not in Python 2.x!

range

In [33]:
list(range(10)) # Let's see what's in the range.  Note it starts at zero!  Where does it stop?

[0, 1, 2, 3, 4, 5, 6, 7, 8, 9]

A more complicated two-input form of the range command produces a range of integers **starting at** a given number, and **terminating before** another given number.

In [37]:
list(range(3,10))

[3, 4, 5, 6, 7, 8, 9]

In [38]:
list(range(-4,5))

[-4, -3, -2, -1, 0, 1, 2, 3, 4]

This is a common source of difficulty for Python beginners.  While the first parameter (-4) is the starting point of the list, the list ends just before the second parameter (5).  This takes some getting used to, but experienced Python programmers grow to like this convention.

The *length* of a list can be accessed by the len command.

In [39]:
len([2,4,6])

3

In [40]:
len(range(10))  # The len command can deal with lists and ranges.  No need to convert.

10

In [41]:
len(range(10,100)) # Can you figure out the length, before evaluating?

90

The final variant of the range command (for now) is the *three-parameter* command of the form `range(a,b,s)`.  This produces a list like `range(a,b)`, but with a "step size" of `s`.  In other words, it produces a list of integers, beginning at `a`, increasing by `s` from one entry to the next, and going up to (but not including) `b`.  It is best to experiment a bit to get the feel for it!

In [42]:
list(range(1,10,2))

[1, 3, 5, 7, 9]

In [43]:
list(range(11,30,2))

[11, 13, 15, 17, 19, 21, 23, 25, 27, 29]

In [44]:
list(range(-4,5,3))

[-4, -1, 2]

In [45]:
list(range(10,100,17))

[10, 27, 44, 61, 78, 95]

This can be used for descending ranges too, and observe that the final number b in range(a,b,s) is not included.

In [46]:
list(range(10,0,-1))

[10, 9, 8, 7, 6, 5, 4, 3, 2, 1]

How many multiples of 7 are between 10 and 100?  We can find out pretty quickly with the range command and the len command (to count).

In [47]:
list(range(10,100,7))  # What list will this create?  It won't answer the question...

[10, 17, 24, 31, 38, 45, 52, 59, 66, 73, 80, 87, 94]

In [48]:
list(range(14,100,7))  # Starting at 14 gives the multiples of 7.

[14, 21, 28, 35, 42, 49, 56, 63, 70, 77, 84, 91, 98]

In [49]:
len(range(14,100,7))  # Gives the length of the list, and answers the question!

13

### Exercises

1.  If `a` and `b` are integers, what is the length of `range(a,b)`?  Express your answer as a formula involving `a` and `b`.

2.  Use a list and range command to produce the list `[1,2,3,4,5,6,7,8,9,10]`.

3.  Create the list [1,2,3,4,5,1,2,3,4,5,1,2,3,4,5,1,2,3,4,5,1,2,3,4,5] with a single list and range command and another operation.

4.  How many multiples of 3 are there between 300 and 3000?

1. Following some experiementation, the length of range(a,b) is b-a if b > a. If b <= a, the range is empty.


2. `list(range(1,11))`


3. `5*list(range(1,6))`


4. Including both 300 and 3000, there are 901 multiples of 3 in this range. This is `len(range(300, 3001, 3))`.

In [53]:
#  Use this space to work on the exercises.
list(range(4,5))

[4]

In [54]:
len(range(300, 3001, 3))

901

<a id='iterating'></a>

## Iterating over a range

Computers are excellent at repetitive reliable tasks.  If we wish to perform a similar computation, many times over, a computer a great tool.  Here we look at a common and simple way to carry out a repetetive computation:  the "for loop".  The "for loop" *iterates* through items in a list or range, carrying out some action for each item.  Two examples will illustrate.

In [56]:
for n in [1,2,3,4,5]:
    print(n*n)

1
4
9
16
25


In [57]:
for s in ['I','Am','Python']:
    print(s + "!")

I!
Am!
Python!


The first loop, **unraveled**, carries out the following sequence of commands.

In [58]:
n = 1
print(n*n)
n = 2
print(n*n)
n = 3
print(n*n)
n = 4
print(n*n)
n = 5
print(n*n)

1
4
9
16
25


But the "for loop" is more efficient *and* more readable to programmers.  Indeed, it saves the repetition of writing the same command `print n*n` over and over again.  It also makes transparent, from the beginning, the range of values that `n` is assigned to.  

When you read and write "for loops", you should consider how they look unravelled -- that is how Python will carry out the loop.  And when you find yourself faced with a repetetive task, you might consider whether it may be wrapped up in a for loop.

Try to unravel the loop below, and predict the result, before evaluating the code.

In [59]:
P = 1
for n in range(1,6):
    P = P * n
print(P)

120


This might have been difficult!  So what if you want to trace through the loop, as it goes?  Sometimes, especially when debugging, it's useful to inspect every step of the loop to see what Python is doing.  We can inspect the loop above, by inserting a print command within the *scope* of the loop.

In [60]:
P = 1
for n in range(1,6):
    P = P * n
    print("n is",n,"and P is",P)
print(P)

n is 1 and P is 1
n is 2 and P is 2
n is 3 and P is 6
n is 4 and P is 24
n is 5 and P is 120
120


Here we have used the *print* command with strings and numbers together.  In Python 3.x, you can print multiple things on the same line by separating them by commas.  The "things" can be strings (enclosed by single or double-quotes) and numbers (int, float, etc.).

In [61]:
print("My favorite number is",17)

My favorite number is 17


If we unravel the loop above, the linear sequence of commands interpreted by Python is the following.

In [62]:
P = 1
n = 1
P = P * n
print("n is",n,"and P is",P)
n = 2
P = P * n
print("n is",n,"and P is",P)
n = 3
P = P * n
print("n is",n,"and P is",P)
n = 4
P = P * n
print("n is",n,"and P is",P)
n = 5
P = P * n
print("n is",n,"and P is",P)
print (P)

n is 1 and P is 1
n is 2 and P is 2
n is 3 and P is 6
n is 4 and P is 24
n is 5 and P is 120
120


Let's analyze the loop syntax in more detail.  
```python
P = 1
for n in range(1,6):
    P = P * n  # this command is in the scope of the loop.
    print("n is",n,"and P is",P)  # this command is in the scope of the loop too!
print(P)
```
The "for" command ends with a colon `:`, and the **next two** lines are indented.  The colon and indentation are indicators of **scope**.  The *scope* of the for loop begins after the colon, and includes all indented lines.  The *scope* of the for loop is what is repeated in every step of the loop (in addition to the reassignment of `n`).  

In [63]:
P = 1
for n in range(1,6):
    P = P * n  # this command is in the scope of the loop.
    print("n is",n,"and P is",P)  # this command is in the scope of the loop too!
print(P)

n is 1 and P is 1
n is 2 and P is 2
n is 3 and P is 6
n is 4 and P is 24
n is 5 and P is 120
120


If we change the indentation, it changes the scope of the for loop.  Predict what the following loop will do, by unraveling, before evaluating it.

In [64]:
P = 1
for n in range(1,6):
    P = P * n
print("n is",n,"and P is",P)
print(P)

n is 5 and P is 120
120


Scopes can be nested by nesting indentation.  What do you think the following loop will do?  Can you unravel it?

In [65]:
for x in [1,2,3]:
    for y in ['a', 'b']:
        print(x,y)

1 a
1 b
2 a
2 b
3 a
3 b


How might you create a nested loop which prints `1 a` then `2 a` then `3 a` then `1 b` then `2 b` then `3 b`?  Try it below.

In [68]:
# Insert your loop here.
for x in ['a', 'b']:
    for y in [1,2,3]:
        print(y,x)

1 a
2 a
3 a
1 b
2 b
3 b


Among popular programming languages, Python is particular about indentation.  Other languages indicate scope with open/close braces, for example, and indentation is just a matter of style.  By requiring indentation to indicate scope, Python effectively removes the need for open/close braces, and enforces a readable style.

We have now encountered data types, operations, variables, and loops.  Taken together, these are powerful tools for computation!  Now complete the following exercises for more practice.

## Exercises

1.  Describe how Python interprets division with remainder when the divisor and/or dividend is negative.
2.  What is the remainder when $2^{90}$ is divided by $91$?
3.  How many multiples of 13 are there between 1 and 1000?
4.  How many *odd* multiples of 13 are there between 1 and 1000?
5.  What is the sum of the numbers from 1 to 1000?
6.  What is the sum of the squares, from $1 \cdot 1$ to $1000 \cdot 1000$?

1. This is interesting. This is all just through testing different things, but this is what I observe. Let $a$ and $m$ be any integers. Mathematically, let $r^{+}$ be the positive "natural representative" of $a$ mod $m$. The 'negative' representative $r^{-}$ is then $m-r^{+}$, where we can write that $a \equiv r^{+} \equiv -r^{-}$ (mod $m$). In Python, the following outputs will occure

$$ 
a \; \% \; m \; = \;
\begin{cases} 
      r^{+} & \text{if} \quad a \geq 0 \; \text{and} \; m > 0 \\
      -r^{-} & \text{if} \quad a \geq 0 \; \text{and} \; m < 0 \\
      r^{-} & \text{if} \quad a < 0 \; \text{and} \; m > 0 \\
      -r^{+} & \text{if} \quad a < 0 \; \text{and} \; m < 0 \\
   \end{cases}
$$

2. 64


3. 76


4. 38


5. 500500


6. 333833500

### Scratch

In [119]:
8 % 5

3

In [110]:
8 % -5

-2

In [109]:
-8 % 5

2

In [111]:
-8 % -5

-3

In [121]:
-13 % 6

5

In [122]:
-13 % -6

-1

In [112]:
2**90 % 91

64

In [113]:
len(range(13, 1000, 13))

76

In [116]:
len(range(13, 1000, 26))

38

In [118]:
S = 0
for n in range(1, 1001):
    S = S + n**2
    
S

333833500

<a id='explore'></a>

# Explorations

Now that you have learned the basics of computation in Pytho and loops, we can start exploring some interesting mathematics!  We are going to look at approximation here -- some ancient questions made easier with programming.

## Exploration 1:  Approximating square roots.

We have seen how Python can do basic arithmetic -- addition, subtraction, multiplication, and division.  But what about other functions, like the square root?  In fact, Python offers a few functions for the square root, but that's not the point.  How can we compute the square root using only basic arithmetic?  

Why might we care?

1.  We might want to know the square root of a number with more precision than the Python function offers.
2.  We might want to understand how the square root is computed... under the hood.
3.  Understanding approximations of square roots and other functions is important, because we might want to approximate other functions in the future (that aren't pre-programmed for us).

Here is a method for approximating the square root of a number X.

1.  Begin with a guess g.
2.  Observe that g * (X / g) = X.  Therefore, among the two numbers g and (X/g), one will be less than or equal to the square root of X, and the other will be greater than or equal to the square root.
3.  Take the average of g and (X/g).  This will be closer to the square root than g or X/g (unless your guess is exactly right!) 
4.  Use this average as a new guess... and go back to the beginning.

Now implement this in Python to approximate the square root of 2.  Use a loop, so that you can go through the approximation process 10 times or 100 times or however many you wish.  Explore the effect of different starting guesses.  Would a change in the averaging function improve the approximation?  How quickly does this converge?  How does this change if you try square roots of different positive numbers?

Write your code (Python) and findings (in Markdown cells) in a readable form.  Answer the questions in complete sentences.



$$\text{Goal:} \quad \sqrt{2} \: = \: 1.4142135623730950488016887242096980785696718753769480731766797379907324784621 \dots$$

In [154]:
root = 2
guess = 1
steps = 10

for i in range(0, steps):
    avg = 0.5*(guess + root/guess)
    guess = avg
    print(guess)
    
print(guess)

1.5
1.4166666666666665
1.4142156862745097
1.4142135623746899
1.414213562373095
1.414213562373095
1.414213562373095
1.414213562373095
1.414213562373095
1.414213562373095
1.414213562373095


### Comments

The code above forms the sequence $\{g_i\}$ given by $$g_i = \frac{g_{i-1}^2 + 2}{2g_{i-1}}$$

The numbers above sugest that this is monotone decreasing, but I have no idea why. I aslo don't know why this converges to $\sqrt{2}$. I believe this follows a different method than what's described below:



Another approach uses the fact that the function $f(x) = x^2$ is increasing: Start with an interval $[g, \ \frac{x}{g}]$ that includes $\sqrt{x}$. Bisect this interval and test the midpoint $m = \frac{g^2 + x}{2g}$. If $m^2 > x$, we know that $m$ is an overestimate. Likewise if $m^2 < x$, then $m$ is an underestimate. In each case, refine your interval putting m as either the inital point or end point, so that the new interval again must contain $\sqrt{x}$. Repeat this process and eventually settle on, say, the midpoint of your last interval as the final approximation. If $D = | \ \frac{x}{g} - g \ |$ is the length of the original interval, then the $n$th interval will have length $\frac{D}{2^n}$. By construction, $\sqrt{x}$ will be within only $\frac{D}{2^n}$ of our chosen midpoint in step $n$. Because these lengths go to zero, the formula can be used to compute $\sqrt{x}$ to any precision, by "solving" the error term, say, for n. I think this is also the way you show that the sequence defined by the midpoints actually converges to some real number as well, though with more details.

## Exploration 2:  Approximating e and pi.

Now we approximate two of the most important constants in mathematics:  e and pi.  There are multiple approaches, but e is pretty easy with the series expansion of e^x.  First, approximate e by the series expansion of e^x at x=1.  How many terms are necessary before the float stabilizes?  Use a loop, with a running product for the factorials and running sums for the series.

$$\text{Goal:} \quad e \: = \: 2.718281828459045235360287471352662497757247093699959574966967627724076630353 \dots$$

In [168]:
steps = 20 # Because we include the constant term in S=1 below, this will approximate e with the taylor polynomial of degree (steps)
N = 1 # For the running factorial
S = 1 # For the running sum. For convenience of the loop, we start at 1 for the 0th degree constant term 1/(0!) = 1.

# Note that the indicies are shifted by 1. This then starts at the linear term 1/(1!).
for k in range(1, steps+1):
    N = N*k
    S += 1/N
    
    print("Step", k, ":  ", S) # Because of the offset, step number will also b the degree of the taylor polynomial used!

print("\nFinal guess:", S)

Step 1 :   2.0
Step 2 :   2.5
Step 3 :   2.6666666666666665
Step 4 :   2.708333333333333
Step 5 :   2.7166666666666663
Step 6 :   2.7180555555555554
Step 7 :   2.7182539682539684
Step 8 :   2.71827876984127
Step 9 :   2.7182815255731922
Step 10 :   2.7182818011463845
Step 11 :   2.718281826198493
Step 12 :   2.7182818282861687
Step 13 :   2.7182818284467594
Step 14 :   2.71828182845823
Step 15 :   2.718281828458995
Step 16 :   2.718281828459043
Step 17 :   2.7182818284590455
Step 18 :   2.7182818284590455
Step 19 :   2.7182818284590455
Step 20 :   2.7182818284590455

Final guess: 2.7182818284590455


### Comments

The float "stabalizes" at step 17, but no ammount of steps gets our approximation over the end `appx` = 2.71...90455. The actual number has $e$ = 2.71...904523.. These do not agree in their final digit... nor even on the rounding of it!

Next we will approximate pi, which is much more interesting.  We can try a few approaches.  For a series-approach (like e), we need a series that converges to pi.  A simple example is the arctangent atan(x).  Recall (precalculus!) that atan(1) = pi/4.  Moreover, the derivative of atan(x) is 1 / (1+x^2).  

1.  Figure out the Taylor series of 1 / (1+x^2) near x=0.  Note that this is a geometric series! 

2.  Figure out the Taylor series of atan(x) near x=0 by taking the antiderivative, term by term, of the above.

3.  Try to estimate pi with this series, using many terms of the series.

$$\text{Goal:} \quad \pi \: = \: 3.1415926535897932384626433832795028841971693993751058209749445923078 \dots$$

The taylor series for arctan(x) is...
$$
\text{arctan}(x) \, \approx \, x - \frac{x^3}{3} + \frac{x^5}{5} - \frac{x^7}{7} + \frac{x^9}{9} - \frac{x^{11}}{11} \pm \cdots 
$$

In [211]:
def atan(x, n):
    # Working from 0 to n-1 uses the 2(n-1) + 1 = (2n - 1)th degree taylor polynomial of arctan(x)
    # I still don't know where this technically converges, buutttt the domain of arctan(x) is all of R
    
    S = 0
    for k in range(0, n):
        S += ((-1)**k) * (1/(2*k+1)) * (x**(2*k+1)) # sign * fraction * x term
        
    return S

In [212]:
# A pretty trash approximation
4*atan(1, 100)

3.1315929035585537

Now we'll accelerate things a bit.  There's a famous formula of Machin (1706) who computed the first hundred digits of pi.  We'll use his identity:

pi/4 = 4 * atan(1/5) - atan(1 / 239).  

This isn't obvious, but there's a tedious proof using sum/difference identities in trig.

Try using this formula now to approximate pi, using your Taylor series for atan(x).  It should require fewer terms.

In [205]:
#  Approximate pi more quickly here!
16*atan(1/5, 11) - 4*atan(1/239, 11)

3.141592653589794

In [206]:
# This doesn't overcome the error at the end (says 4 instead of 3), at all it seems
# As above though, it his this only at n=11 steps
16*atan(1/5, 100) - 4*atan(1/239, 100)

3.141592653589794

Now let's compare this to **Archimedes' method**.  Archimedes approximated pi by looking at the perimeters p(n) and P(n) of (2^n)-gons inscribed in and circumscribed around a unit circle.  So p(2) is the perimeter of a square inscribed in the unit circle.  P(2) is the perimeter of a square circumscribed around a unit circle.

Archimedes proved the following (not in the formulaic language of algebra):  For all n >= 2,

(P-formula)  P(n+1) = 2 * p(n) * P(n) / (p(n) + P(n)).

(p-formula)  p(n+1) = sqrt( p(n) * P(n+1) ).

1.  Compute p(2) and P(2).

2.  Use these formulas to compute p(10) and P(10).  Use this to get a good approximation for pi!

We could use our previous sqrt function if you want, we'll take a fancier high-precision approach.  "mpmath" is a Python package for high-precision calculation.  It should come with your Anaconda installation.  You can read the full documentation at http://mpmath.org/doc/current/

First we load the package and print its status.

In [207]:
from mpmath import *
print(mp)

Mpmath settings:
  mp.prec = 53                [default: 53]
  mp.dps = 15                 [default: 15]
  mp.trap_complex = False     [default: False]


The number mp.dps is (roughly) the number of decimal digits that mpmath will keep track of in its computations.  mp.prec is the binary precision, a bit more than 3 times the decimal precision.  We can change this to whatever we want.

In [213]:
mp.dps = 50 # Let's try 50 digits precision to start.
print(mp)

Mpmath settings:
  mp.prec = 169               [default: 53]
  mp.dps = 50                 [default: 15]
  mp.trap_complex = False     [default: False]


mpmath has a nice function for square roots.  Compare this to your approximation from before!

In [214]:
sqrt(2)  # mpf(...) stands for an mp-float.  

mpf('1.4142135623730950488016887242096980785696718753769468')

In [215]:
type(sqrt(2)) # mpmath stores numbers in its own types!

mpmath.ctx_mp_python.mpf

In [216]:
4*mp.atan(1) # mpmath has the arctan built in.  This should be pretty close to pi!

mpf('3.1415926535897932384626433832795028841971693993751068')

Now try Archimedes' approximation of pi.  Use the mpmath sqrt function along the way.  How many iterations do you need to get pi correct to 100 digits?  Compare this to the arctan-series (not the mpmath atan function) via Machin's formula.

## Lesson: DON'T do the following

In [241]:
#  Explore and experiment.
def little_p(n):
    
    # To make sure we're not dying
    print("Little p step", n)
    
    if n < 2 or type(n) != int:
        print("n has to be an integer greater than or equal to 2!!")
        return
    
    elif n == 2:
        return 4*sqrt(2)
    
    else: 
        val = sqrt(little_p(n-1) * big_p(n))
        
    return val



def big_p(n):
    
    if n < 2 or type(n) != int:
        print("n has to be an integer greater than or equal to 2!!")
        return
    
    elif n == 2:
        return 8
    
    else: 
        val = (2 * little_p(n-1) * big_p(n-1)) / (little_p(n-1) + big_p(n-1))
        
    return val   


In [231]:
little_p(2)

mpf('5.6568542494923801952067548968387923142786875015077873')

In [232]:
big_p(2)

8

In [ ]:
# This took like ages, removed the output though
#  because it made my pdf 700 pages from the "Little p step" check I put.
#  Even with the input n=9, the function ran like a million times. Horrible idea!
pi_appx = 0.5*little_p(9)
pi_appx

## This should be better:

In [287]:
# Given a specified n, the following code generates two lists containing
# the sequences {p(i)} and {P(i)} for 2 <= i <= n. Notice that little_p(n+1)
# depends on big_p(n+1). This is taken in account in the order of operations done in the loop.

# Because our lists are already initialized at n=2, we need only
# perform (n - 2) additional itterations to hit p(n).
n = 10000
steps = n - 2

# Initialize the arrays of our sequence, they hold p(2) and P(2) respectively
lil_p_list = [4*sqrt(2)]
big_p_list = [8]

# As mentioned, i=0 here corresponds with n=2 for p(n), in particular, n=i+2.
for i in range(0, steps):
    
    # Grab and set p
    p_n = lil_p_list[i]   # p(n)
    PP_n = big_p_list[i]  # P(n), look how the capital P actually looks smaller than the lowercase p. P vs p
    
    # Construct P(n+1) and add it to the list
    PP_next = (2*p_n*PP_n)/(p_n + PP_n)
    big_p_list.append(PP_next)
    
    # Construct p(n+1) and add it it to the list
    p_next = sqrt(p_n * PP_next)
    lil_p_list.append(p_next)

# Print results:
print("Results:")
print("little p("+str(n)+") = " + str(lil_p_list[steps]))  # Following ALL the indicies, these ARE big P(n)
print("big P("+str(n)+") = "    + str(big_p_list[steps]))  # and little p(n), finally.

# Since big P is always an over estimate, and little p an
# under lets be really cool and say...
approximation = 0.5*0.5*(lil_p_list[steps] + big_p_list[steps])

print("\nApproximation:", approximation)

Results:
little p(10000) = 6.2831853071795864769252867665590057683943387987502
big P(10000) = 6.2831853071795864769252867665590057683943387987502

Approximation: 3.1415926535897932384626433832795028841971693993751


### Comment
This seems much much better than the atan approximation!